In [1]:
# import library
import numpy as np 
import pandas as pd

# visualization
import matplotlib.pyplot as plt 
import seaborn as sns 

# menampilkan semua kolom
pd.set_option('display.max_columns', None)

# menghilangkan warning
import warnings
warnings.filterwarnings('ignore')

In [2]:
#pd.set_option('display.max_rows', 140)

# **1. Introduction** 


# **Business Understanding**

### **Context**

Transjakarta is a Bus Rapid Transit (BRT) system serving the city of Jakarta, Indonesia. Launched in 2004, it was the first BRT system in Southeast Asia and South Asia. Transjakarta operates a network of bus lines throughout the city, using dedicated bus lanes to provide faster, more efficient public transportation. The system aims to reduce traffic congestion, improve air quality, and provide affordable transportation options for Jakarta's residents and visitors. Transjakarta uses a tap-in, tap-out system with smart cards for fare collection, allowing for detailed tracking of passenger journeys and system usage.


### **Problem statement**

As Jakarta continues to grow and urbanize, efficient public transportation becomes increasingly crucial. Transjakarta plays a vital role in moving millions of people daily, but faces challenges in optimizing its service to meet the evolving needs of the city. With the availability of detailed transaction data, there's an opportunity to gain insights that can lead to significant improvements in service quality, operational efficiency, and passenger satisfaction.


### **Challenges and Opportunities for New Airbnb Partners**

While Transjakarta has been successful in providing a much-needed transportation solution, it faces several challenges:

**Route Performance**

Challenge: Identifying the most and least used Transjakarta routes.
Goal: Optimize resource allocation by potentially increasing services on popular routes and reviewing the necessity of less-used routes.


**Stop Utilization**

Challenge: Understanding which Transjakarta stops are most frequently used.
Goal: Improve infrastructure and services at high-traffic stops to enhance passenger experience.


**Demographic Analysis**

Challenge: Analyzing usage patterns based on gender and age (birth decade).
Goal: Tailor services and marketing strategies to cater to the primary user demographics.


**Payment System Efficiency**

Challenge: Identifying the most popular bank cards and analyzing fare distribution.
Goal: Streamline payment processes and potentially negotiate better terms with frequently used banks.


**Temporal Usage Patterns**

Challenge: Understanding usage patterns throughout the day and peak hours.
Goal: Adjust service frequency to match demand during different times of the day.


**Journey Duration Analysis**

Challenge: Analyzing the duration of passenger journeys.
Goal: Identify opportunities to reduce travel times and improve service efficiency.




### **Goal: Guiding New Partners for Success**

Through this data analysis project, we aim to answer the following question:

**What key recommendations can be provided to Transjakarta to improve its service quality, operational efficiency, and meet the evolving transportation needs of Jakarta's residents?
**

By providing data-driven recommendations, we can help Transjakarta optimize its operations, enhance the passenger experience, and contribute to a more sustainable and efficient transportation system for Jakarta. This, in turn, will support the city's goals for reduced congestion, improved air quality, and enhanced mobility for its citizens.

# **2. Data Preparation**

Before beginning the analysis, we need to carry out data understanding and data cleaning simultaneously. The data understanding phase helps us explore the dataset's characteristics, such as identifying unique values, data types, duplicates, missing values, outliers, and anomalies. As these issues are recognized, we immediately address them in the data cleaning phase, which involves handling duplicates, missing values, and anomalies. I won't seperate these into 2 different subpart because i'll clean the data straight after i understand where the data is lacking. So both the Data Understanding and Data Cleaning will both be called Data Preparation

## **Dataset and libraries**

There are 22 columns. The description of each column from Transjakarta.csv dataset can be seen as below

|Column name | Descriptions|
|---|------------------------|
|**transID**     | Unique transaction id for every transaction|
|**payCardID**   | Customers main identifier. The card customers use as a ticket for entrance and exit.|
|**payCardBank**| Customers card bank issuer name|
|**payCardName**| Customers name that is embedded in the card. |
|**payCardSex**| Customers sex that is embedded in the card|
|**payCardBirthDate**| Customers birth year|
|**corridorID**| Corridor ID / Route ID as key for route grouping. |
|**corridorName**| Corridor Name / Route Name contains Start and Finish for each route. |
|**direction**| 0 for Go, 1 for Back. Direction of the route. |
|**tapInStops**| Tap In (entrance) Stops ID for identifying stops name |
|**tapInStopsName**| Tap In (entrance) Stops Name where customers tap in. |
|**tapInStopsLat**| Latitude of Tap In Stops.|
|**tapInStopsLon**| Longitude of Tap In Stops.|
|**stopStartSeq**| Sequence of the stops, 1st stop, 2nd stops etc. Related to direction..|
|**tapInTime**| Time of tap in. Date and time|
|**tapOutStops**     | Tap Out (Exit) Stops ID for identifying stops name|
|**tapOutStopsName**   | Tap out (exit) Stops Name where customers tap out|
|**tapOutStopsLat**| Latitude of Tap Out Stops|
|**tapOutStopsLon**| Longitude of Tap Out Stops |
|**stopEndSeq**   | Sequence of the stops, 1st stop, 2nd stops etc. Related to direction.|
|**tapOutTime**| Time of tap out. Date and time|
|**payAmount**| The number of what customers pay. Some are free. Some not.|


In [3]:
transjakarta = pd.read_csv('Transjakarta.csv')
transjakarta

transID            payCardID payCardBank  \
0      EIIW227B8L34VB      180062659848800      emoney   
1      LGXO740D2N47GZ     4885331907664776         dki   
2      DJWR385V2U57TO     4996225095064169         dki   
3      JTUZ800U7C86EH         639099174703       flazz   
4      VMLO535V7F95NJ         570928206772       flazz   
...               ...                  ...         ...   
37895  ZWEC949B8Q87QG  4685818286724028395      brizzi   
37896  YHHK837P6Y95GN     6502902290603767         dki   
37897  YXPP627N4G95HO      213159426675861      emoney   
37898  RGVK175U2U98UV      377840859133591      emoney   
37899  FMZZ963S4B68ZP         501862539795       flazz   

                  payCardName payCardSex  payCardBirthDate corridorID  \
0              Bajragin Usada          M              2008          5   
1                Gandi Widodo          F              1997         6C   
2               Emong Wastuti          F              1992        R1A   
3                Surya Wacana          F              1978        11D   
4             Embuh Mardhiyah          M              1982         12   
...                       ...        ...               ...        ...   
37895         Kamila Mahendra          F              2004         6B   
37896            Titi Siregar          M              1974         9N   
37897  drg. Zahra Nashiruddin          F              1976         1T   
37898            Ana Agustina          M              1976     JAK.13   
37899       drg. Leo Najmudin          F              1985         13   

                                   corridorName  direction tapInStops  \
0                         Matraman Baru - Ancol        1.0     P00142   
1      Stasiun Tebet - Karet via Patra Kuningan        0.0    B01963P   
2                            Pantai Maju - Kota        0.0    B00499P   
3           Pulo Gebang - Pulo Gadung 2 via PIK        0.0    B05587P   
4                         Tanjung Priok - Pluit        0.0     P00239   
...                                         ...        ...        ...   
37895         Ragunan - MH Thamrin via Semanggi        1.0     P00261   
37896                    Pinang Ranti - Pramuka        1.0     P00064   
37897                      Cibubur - Balai Kota        1.0    B02873P   
37898               Tanah Abang - Jembatan Lima        1.0    B02505P   
37899                         Ciledug - Tendean        0.0     P00001   

               tapInStopsName  tapInStopsLat  tapInStopsLon  stopStartSeq  \
0                   Pal Putih      -6.184631      106.84402             7   
1                  Kemenkes 2      -6.228700      106.83302            13   
2                Gg. Kunir II      -6.133132      106.81435            38   
3                Taman Elok 1      -6.195743      106.93526            23   
4      Sunter Boulevard Barat      -6.149650      106.88900             5   
...                       ...            ...            ...           ...   
37895                  Tosari      -6.196892      106.82309             2   
37896       Garuda Taman Mini      -6.290154      106.88116             1   
37897           Plaza Sentral      -6.216247      106.81676            12   
37898          Museum Textile      -6.188656      106.80954            33   
37899              Adam Malik      -6.236466      106.74786             2   

                 tapInTime tapOutStops        tapOutStopsName  tapOutStopsLat  \
0      2023-04-03 05:21:44      P00253                Tegalan       -6.203101   
1      2023-04-03 05:42:44     B03307P    Sampoerna Strategic       -6.217152   
2      2023-04-03 05:59:06     B04962P  Simpang Kunir Kemukus       -6.133731   
3      2023-04-03 05:44:51     B03090P      Raya Penggilingan       -6.183068   
4      2023-04-03 06:17:35      P00098       Kali Besar Barat       -6.135355   
...                    ...         ...                    ...             ...   
37895  2023-04-21 18:18:37      P00228                 SMK 57       -6.

In [4]:
# Displaying the summary of the dataset
transjakarta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37900 entries, 0 to 37899
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   transID           37900 non-null  object 
 1   payCardID         37900 non-null  int64  
 2   payCardBank       37900 non-null  object 
 3   payCardName       37900 non-null  object 
 4   payCardSex        37900 non-null  object 
 5   payCardBirthDate  37900 non-null  int64  
 6   corridorID        36643 non-null  object 
 7   corridorName      35970 non-null  object 
 8   direction         37900 non-null  float64
 9   tapInStops        36687 non-null  object 
 10  tapInStopsName    37900 non-null  object 
 11  tapInStopsLat     37900 non-null  float64
 12  tapInStopsLon     37900 non-null  float64
 13  stopStartSeq      37900 non-null  int64  
 14  tapInTime         37900 non-null  object 
 15  tapOutStops       35611 non-null  object 
 16  tapOutStopsName   36556 non-null  object

In [5]:
# Checking the total of NA in our data
transjakarta.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID          1257
corridorName        1930
direction              0
tapInStops          1213
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops         2289
tapOutStopsName     1344
tapOutStopsLat      1344
tapOutStopsLon      1344
stopEndSeq          1344
tapOutTime          1344
payAmount           1007
dtype: int64

In [6]:
# Checking dupilicated data
transjakarta.duplicated().sum()

np.int64(0)

We will be making a table with complete dtype, unique, and nunique of each columns so that we can get a generalized view of our data

In [7]:
# Shortening the name for faster and easier code writing
tj = transjakarta

columns = [i for i in tj.columns]
nunique = [tj[i].nunique() for i in tj.columns]
dtype = [tj[i].dtype for i in tj.columns]
unique = [tj[i].unique() for i in tj.columns]

pd.set_option('display.max_colwidth', 300)
data_values = pd.DataFrame({
    'columns' : columns,
    'data type' : dtype,
    'nunique' : nunique,
    'unique' : unique
})

data_values

columns data type  nunique  \
0            transID    object    37900   
1          payCardID     int64     2000   
2        payCardBank    object        6   
3        payCardName    object     1993   
4         payCardSex    object        2   
5   payCardBirthDate     int64       67   
6         corridorID    object      221   
7       corridorName    object      216   
8          direction   float64        2   
9         tapInStops    object     2570   
10    tapInStopsName    object     2602   
11     tapInStopsLat   float64     2587   
12     tapInStopsLon   float64     2458   
13      stopStartSeq     int64       67   
14         tapInTime    object    37079   
15       tapOutStops    object     2230   
16   tapOutStopsName    object     2248   
17    tapOutStopsLat   float64     2237   
18    tapOutStopsLon   float64     2139   
19        stopEndSeq   float64       74   
20        tapOutTime    object    35908   
21         payAmount   float64        3   

                                                                                                                                                                                                                                                                                                         unique  
0   [EIIW227B8L34VB, LGXO740D2N47GZ, DJWR385V2U57TO, JTUZ800U7C86EH, VMLO535V7F95NJ, DDES630K2F80KC, HEMW326B9N91TV, XTKE052E5E87LN, OIHS248V7S72EB, ERXO372B2H63RB, CLWK627W2M22WO, IIEK980X1O23NZ, OPLW560U3S06UG, XESX233R4U46DN, LZFD368Q3E81HJ, KLUN542I2F21PI, XDMP644Z5Y29DB, CUPO258C0D42PY, SBPQ144...  
1   [180062659848800, 4885331907664776, 4996225095064169, 639099174703, 570928206772, 2251412124634980, 213155822653833, 3587341902618993, 6510013988638519, 3580401035990896, 5541851157087967, 347204308990092, 3536759468239783, 6575130739600630, 4056836188316, 4532407180860336, 4556326832434362836, ...  
2                                                                                                                                                                                                                                                                     [emoney, dki, flazz, online, brizzi, bni]  
3   [Bajragin Usada, Gandi Widodo, Emong Wastuti, Surya Wacana, Embuh Mardhiyah, Tirta Siregar, R. Devi Hariyah, S.T., Darmanto Rajasa, Cagak Maheswara, S.Sos, T. Dadap Pradana, M.Ak, Oskar Widiastuti, S.Farm, R.M. Labuh Laksmiwati, Shakila Maryadi, Danuja Usada, Nadine Utami, H. Lutfan Nurdiyanti, ...  
4                                                                                                                                                                                                                                                                                                        [M, F]  
5   [2008, 1997, 1992, 1978, 1982, 1993, 1974, 1991, 2004, 1984, 1988, 2002, 1975, 2011, 2003, 1981, 2009, 1983, 1999, 2007, 1979, 1960, 1995, 2001, 1990, 1968, 1985, 2010, 1989, 2006, 1986, 1965, 2005, 1966, 1994, 1972, 1969, 1970, 2000, 1953, 1977, 2012, 1987, 1996, 1971, 1956, 1980, 1998, 1973, 1...  
6   [5, 6C, R1A, 11D, 12, 1T, JAK.18, nan, B14, 1Q, 5M, M7B, 9A, 4, JAK.46, S22, 13, 9E, M9, JAK.88, T11, JAK.16, 6V, 11Q, JAK.02, 11, 1C, 3B, M1, 7E, 1P, 7D, 6H, 14, BW9, JIS3, 2, JAK.15, 6N, M11, JAK.112, 9N, JAK.43B, JAK.42, 7P, JAK.36, 13B, JAK.06, JAK.14, M7, T21, 12A, 9D, 1, 3A, 3C, M8, B21, J...  
7   [Matraman Baru - Ancol, Stasiun Tebet - Karet via Patra Kuningan, Pantai Maju - Kota, Pulo Gebang - Pulo Gadung 2 via PIK, Tanjung Priok - Pluit, Cibubur - Balai Kota, Kalibata - Kuningan, nan, Bekasi Barat - Kuningan, Rempoa - Blok M, Kampung Melayu - Tanah Abang via Cikini, BKN - Blok M, PGC 2...  
8                                                                                                                                                                                                                                                                            

Based on the observation of unique values from data frame, we notice some finding such as,
* There's only 2000 unique payCardID which means there's only 2000 unique cusomer from 37900 Transjakarta usage. This shows a huge reccuring usage from the cusomer in our data
* There is more than 200 corridor ID which means there is more than 200 transjakarta route available
* There should be equal number of corridorID and corridorName because each corridorName represents unique corridorID
* There's more than 2500 tapInStops or Halte, with 220 route that make the number of average stops per route of 12
* stopStartSeq represents Halte number, which means each stops have their own unique seqence number for their own route, based on the previous facts, this sequence is surprising because theres a route with more than 60 stops, which seems excessive and we'll be checking if the data is true
* There is 4 different price which is wierd considering all Transjakarta usage should have the same price of 3500
* We will also be changing the stopEndSeq to match the stopStartSeq which is an integer because every sequence must be an integer


In [8]:
# cheking the total number of sequence of each route
pd.pivot_table(tj,
               index = 'corridorID',
               values = 'stopStartSeq',
               aggfunc='max').sort_values('stopStartSeq', ascending= False).head(20)

stopStartSeq
corridorID              
JAK.33                68
JAK.53                68
JAK.58                66
JAK.30                65
JAK.110A              64
JAK.56                61
JAK.32                61
JAK.18                58
JAK.72                58
JAK.31                56
JAK.50                56
JAK.12                55
JAK.19                53
JAK.14                53
JAK.49                52
JAK.17                51
5N                    50
JAK.28                50
JAK.29                50
JAK.60                49

after crosschecking with transjakarta official website, transjakarta.co.id, turns out these route truly have such a high number of stops or sequence

Now We will use the corridorID to fill in the missing values in corridorName and vice versa. Before doing that, we need to ensure that the ID and Name are correct. corridorID and corridorName should have the same number of unique values because they are paired: a single corridorID represents a single corridorName. However, we found that there are 221 corridorIDs and only 216 corridorNames. We need to clean these values.

## **MAKING DataFrame for LOOKUP condition : data_stasiun_count**

we will be using something functionally similar to a VLOOKUP or HLOOKUP from excel, so I'll be calling these pivot table where we match these table to the tj dataframe a "LOOKUP dataframe". This LOOKUP will be used for corridor Name and ID, If the ID is NA, then we will match the Name from TJ and LOOKUP and then fill the ID when we found the matching Name

In [9]:
# Making a pivot table, since the corridorName has lower count, there should be a corridorName with multiple corridorID
data_stasiun_count = pd.pivot_table(data = tj,
               index = ['corridorName','corridorID'],
               values = 'tapInStopsLat',
               aggfunc= 'count').sort_values(by=['tapInStopsLat'], ascending=False).reset_index().rename(columns={'tapInStopsLat': 'Used'})

data_stasiun_count

corridorName corridorID  Used
0                              Cibubur - Balai Kota         1T   391
1                                     Ciputat - CSW        S21   383
2           Harmoni - Jakarta International Stadium       JIS3   337
3                      Kebayoran Lama - Tanah Abang         8C   333
4                    Kampung Rambutan - Pondok Gede     JAK.06   329
..                                              ...        ...   ...
216                         Gondangdia - Balai Kota         2Q    40
217                     Term. Pulo Gadung - Lampiri     JAK.99    31
218  Tanah Abang - Kebayoran Lama via Pos Pengumben     JAK.12    23
219                       Kampung Rambutan - Blok M         7B    16
220                           Matraman Baru - Ancol         M5    15

[221 rows x 3 columns]

We will also be checking our previous questions of:
* There should be equal number of corridorID and corridorName because each corridorName represents unique corridorID

by looking at the duplicates

In [10]:
# Identify the corridorName values with more than one corridorID
duplicated_corridor = data_stasiun_count.groupby('corridorName')['corridorID'].count().reset_index()
duplicated_corridor = duplicated_corridor[duplicated_corridor['corridorID'] > 1]

# Filter the original DataFrame to get only the rows with duplicated corridorName values
data_stasiun_count[data_stasiun_count['corridorName'].isin(duplicated_corridor['corridorName'])]

corridorName corridorID  Used
14                   Pulo Gadung - Monas          2   279
26                 Matraman Baru - Ancol          5   246
39                         Blok M - Kota          1   230
59   Kalideres - Bundaran HI via Veteran         M3   204
114                 Pinang Ranti - Pluit         M9   149
152  Kalideres - Bundaran HI via Veteran          3   127
202                        Blok M - Kota         M1    61
205                  Pulo Gadung - Monas         M2    60
207                 Pinang Ranti - Pluit          9    59
220                Matraman Baru - Ancol         M5    15

After reviewing the Transjakarta Official Website transjakarta.co.id, it turns out that these 5 duplicated corridorIDs are correct. Transjakarta has different corridorIDs for these corridorNames.

## **FILLING CorridorName and CorridorID with LAMBDA**

Now we will be matching the LOOKUP DataFrame with our tj DataFrame

In [11]:
def fill_missing_corridor(source, apply_to):
    # Create a dictionary for fast lookup
    id_to_name = dict(zip(data_stasiun_count[source], data_stasiun_count[apply_to]))
    
    # Fill missing 'corridorName' in tj using the dictionary
    tj[apply_to] = tj.apply(
        lambda row: id_to_name[row[source]] if pd.isna(row[apply_to]) and not pd.isna(row[source]) else row[apply_to],
        axis=1
    )
    
    return tj

fill_missing_corridor('corridorID', 'corridorName')
fill_missing_corridor('corridorName', 'corridorID')


transID            payCardID payCardBank  \
0      EIIW227B8L34VB      180062659848800      emoney   
1      LGXO740D2N47GZ     4885331907664776         dki   
2      DJWR385V2U57TO     4996225095064169         dki   
3      JTUZ800U7C86EH         639099174703       flazz   
4      VMLO535V7F95NJ         570928206772       flazz   
...               ...                  ...         ...   
37895  ZWEC949B8Q87QG  4685818286724028395      brizzi   
37896  YHHK837P6Y95GN     6502902290603767         dki   
37897  YXPP627N4G95HO      213159426675861      emoney   
37898  RGVK175U2U98UV      377840859133591      emoney   
37899  FMZZ963S4B68ZP         501862539795       flazz   

                  payCardName payCardSex  payCardBirthDate corridorID  \
0              Bajragin Usada          M              2008          5   
1                Gandi Widodo          F              1997         6C   
2               Emong Wastuti          F              1992        R1A   
3                Surya Wacana          F              1978        11D   
4             Embuh Mardhiyah          M              1982         12   
...                       ...        ...               ...        ...   
37895         Kamila Mahendra          F              2004         6B   
37896            Titi Siregar          M              1974         9N   
37897  drg. Zahra Nashiruddin          F              1976         1T   
37898            Ana Agustina          M              1976     JAK.13   
37899       drg. Leo Najmudin          F              1985         13   

                                   corridorName  direction tapInStops  \
0                         Matraman Baru - Ancol        1.0     P00142   
1      Stasiun Tebet - Karet via Patra Kuningan        0.0    B01963P   
2                            Pantai Maju - Kota        0.0    B00499P   
3           Pulo Gebang - Pulo Gadung 2 via PIK        0.0    B05587P   
4                         Tanjung Priok - Pluit        0.0     P00239   
...                                         ...        ...        ...   
37895         Ragunan - MH Thamrin via Semanggi        1.0     P00261   
37896                    Pinang Ranti - Pramuka        1.0     P00064   
37897                      Cibubur - Balai Kota        1.0    B02873P   
37898               Tanah Abang - Jembatan Lima        1.0    B02505P   
37899                         Ciledug - Tendean        0.0     P00001   

               tapInStopsName  tapInStopsLat  tapInStopsLon  stopStartSeq  \
0                   Pal Putih      -6.184631      106.84402             7   
1                  Kemenkes 2      -6.228700      106.83302            13   
2                Gg. Kunir II      -6.133132      106.81435            38   
3                Taman Elok 1      -6.195743      106.93526            23   
4      Sunter Boulevard Barat      -6.149650      106.88900             5   
...                       ...            ...            ...           ...   
37895                  Tosari      -6.196892      106.82309             2   
37896       Garuda Taman Mini      -6.290154      106.88116             1   
37897           Plaza Sentral      -6.216247      106.81676            12   
37898          Museum Textile      -6.188656      106.80954            33   
37899              Adam Malik      -6.236466      106.74786             2   

                 tapInTime tapOutStops        tapOutStopsName  tapOutStopsLat  \
0      2023-04-03 05:21:44      P00253                Tegalan       -6.203101   
1      2023-04-03 05:42:44     B03307P    Sampoerna Strategic       -6.217152   
2      2023-04-03 05:59:06     B04962P  Simpang Kunir Kemukus       -6.133731   
3      2023-04-03 05:44:51     B03090P      Raya Penggilingan       -6.183068   
4      2023-04-03 06:17:35      P00098       Kali Besar Barat       -6.135355   
...                    ...         ...                    ...             ...   
37895  2023-04-21 18:18:37      P00228                 SMK 57       -6.

In [12]:
transjakarta.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID          1125
corridorName        1125
direction              0
tapInStops          1213
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops         2289
tapOutStopsName     1344
tapOutStopsLat      1344
tapOutStopsLon      1344
stopEndSeq          1344
tapOutTime          1344
payAmount           1007
dtype: int64

After filling the NA with pivot table, there's still a lot of NA in the CorridorID and CorridorName, looking at it, there's only 221 rows of corridorName and ID combination from our pivot table, but when taking a data from transjakarta.co.id there's a 252 rows of corridorName and ID. In conclusion, there might be a corridorName without corridorID and vice versa from out tj dataframe. Now I'm trying to fill the NA with csv data from transjakarta.co.id

In [13]:
corridor = pd.read_csv('TJ corridor.csv', sep = ':')
corridor

corridorID                               corridorName
0           1                               BLOK M - KOTA
1           2              PULO GADUNG - MONUMEN NASIONAL
2          2A                    PULO GADUNG - RAWA BUAYA
3           3    KALIDERES - MONUMEN NASIONAL VIA VETERAN
4          3F                KALIDERES - SENAYAN BANK DKI
..         ...                                        ...
247        M5                      KAMPUNG MELAYU - ANCOL
248        M6                        RAGUNAN - GALUNGGUNG
249        M7           KAMPUNG RAMBUTAN - KAMPUNG MELAYU
250        M8                    LEBAK BULUS - PASAR BARU
251        M9                        PINANG RANTI - PLUIT

[252 rows x 2 columns]

The corridorName from the transjakarta website use all uppercase letter so we have to do the same to find a match with our LOOKUP dataframe

In [14]:
tj['corridorName'] = tj['corridorName'].fillna(
    tj['corridorID'].str.upper().map(corridor.set_index('corridorID')['corridorName'])
)

In [15]:
transjakarta.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID          1125
corridorName        1125
direction              0
tapInStops          1213
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops         2289
tapOutStopsName     1344
tapOutStopsLat      1344
tapOutStopsLon      1344
stopEndSeq          1344
tapOutTime          1344
payAmount           1007
dtype: int64

In [16]:
corridor.drop_duplicates(subset='corridorName', inplace= True)

In [17]:
tj['corridorID'] = tj['corridorID'].fillna(
    tj['corridorName'].str.upper().map(corridor.set_index('corridorName')['corridorID'])
)

In [18]:
transjakarta.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID          1125
corridorName        1125
direction              0
tapInStops          1213
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops         2289
tapOutStopsName     1344
tapOutStopsLat      1344
tapOutStopsLon      1344
stopEndSeq          1344
tapOutTime          1344
payAmount           1007
dtype: int64

Turns out using outside more complete data doesn really have any effect in filling the NA

## **MAKING DataFrame for LOOKUP condition : data_halte_count**

Now we will be filling the NA from the tapInStops column. First we make the LOOKUP table were we find every combination of stops Name and ID from our tj DataFrame

In [19]:
data_halte_count = pd.pivot_table(data = tj,
               index = ['tapInStops','tapInStopsName'],
               values = 'tapInStopsLat',
               aggfunc= 'count').sort_values(by=['tapInStops']).reset_index().rename(columns={'tapInStopsLat': 'used'}).sort_values('used', ascending = False)

data_halte_count


tapInStops            tapInStopsName  used
2462     P00170               Penjaringan   236
2370     P00064         Garuda Taman Mini   200
2328     P00016                       BKN   170
2568     P00297                   BNN LRT   151
2457     P00164                   Pejaten   145
...         ...                       ...   ...
26      B00059P  Apotik Emilia Rawamangun     1
24      B00056P   Apartemen Kedoya Elok 2     1
23      B00052P                Antena III     1
39      B00090P   Azhari Islamic School 1     1
101     B00206P              Bulak Rantai     1

[2570 rows x 3 columns]

Checking for duplicates because the map function will return an error if there is a duplicated values inside the lookup table

In [20]:
# Check for duplicates in the 'tapInStops' column
data_halte_count[data_halte_count['tapInStops'].duplicated(keep=False)]


Empty DataFrame
Columns: [tapInStops, tapInStopsName, used]
Index: []

In [21]:
# Check for duplicates in the 'tapInStopsName' column
data_halte_count[data_halte_count['tapInStopsName'].duplicated(keep=False)]


tapInStops tapInStopsName  used
2359     P00052   Dukuh Atas 2    20
2569     P00298   Dukuh Atas 2     1

We find the duplicates, This Stops Name should only have 1 ID, and since the one above have been used more, I'll take it as the correct one.

In [22]:
tj[tj['tapInStops'] == 'P00298']

transID     payCardID payCardBank       payCardName payCardSex  \
33176  RVXV189M7J06ND  501834837947       flazz  Cahyanto Maulana          M   

       payCardBirthDate corridorID                  corridorName  direction  \
33176              1985         M4  Pulo Gadung 2 - Dukuh Atas 2        1.0   

      tapInStops tapInStopsName  tapInStopsLat  tapInStopsLon  stopStartSeq  \
33176     P00298   Dukuh Atas 2      -6.204494     106.823327             0   

                 tapInTime tapOutStops tapOutStopsName  tapOutStopsLat  \
33176  2023-04-15 06:55:23      P00152   Pasar Genjing       -6.194446   

       tapOutStopsLon  stopEndSeq           tapOutTime  payAmount  
33176       106.86092         4.0  2023-04-15 07:53:38     3500.0

## **FILLING tapInStopsName with MAP**

In [23]:
# Deleting the duplicate, it will automatically drop the second occurance and since our LOOKUP is sorted we'll drop the one that is used only once
data_halte_count = data_halte_count.drop_duplicates(subset='tapInStopsName')

# Map the 'source' column in tj to the 'apply_to' column in data_stasiun_count and fill missing values
tj['tapInStops'] = tj['tapInStops'].fillna(
    tj['tapInStopsName'].map(data_halte_count.set_index('tapInStopsName')['tapInStops'])
)

In [24]:
transjakarta.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID          1125
corridorName        1125
direction              0
tapInStops            34
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops         2289
tapOutStopsName     1344
tapOutStopsLat      1344
tapOutStopsLon      1344
stopEndSeq          1344
tapOutTime          1344
payAmount           1007
dtype: int64

We successfully fill almost all of the tapInStops, these remaining NA means we don't find a match with our LOOKUP table and we have to either manually fill them, or drop them. Since there is only 34 data out of 37 thousand data, the significance of these 34 data is so low that it is not worth filling the NA manually.

## **FILLING tapOutStops with MAP**

We will be using the same process as the one above, The same LOOKUP table only this time we're using the Stops OUT instead of IN. Then we will use this LOOKUP table to fill the TapOutStopsName

In [25]:
data_halte_count2 = pd.pivot_table(data = tj,
               index = ['tapOutStops','tapOutStopsName'],
               values = 'tapOutStopsLat',
               aggfunc= 'count').sort_values(by=['tapOutStops']).reset_index().rename(columns={'tapOutStopsLat': 'used'}).sort_values('used', ascending = False)

data_halte_count2.drop_duplicates(subset='tapOutStopsName', inplace = True)


In [26]:
tj['tapOutStops'] = tj['tapOutStops'].fillna(
    tj['tapOutStopsName'].map(data_halte_count2.set_index('tapOutStopsName')['tapOutStops']))

In [27]:
transjakarta.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID          1125
corridorName        1125
direction              0
tapInStops            34
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops         1363
tapOutStopsName     1344
tapOutStopsLat      1344
tapOutStopsLon      1344
stopEndSeq          1344
tapOutTime          1344
payAmount           1007
dtype: int64

It occurs to me that the still missing stops ID is because our LOOKUP table doesn't represent all the stops population, only the stops sample from our data which is not complete. So there might be a difference in the tapIn and tapOut stops and we can try Mapping them to each outher LOOKUP table

In [28]:
tj['tapInStops'] = tj['tapInStops'].fillna(
    tj['tapInStopsName'].map(data_halte_count2.set_index('tapOutStopsName')['tapOutStops'])
)

In [29]:
tj['tapOutStops'] = tj['tapOutStops'].fillna(
    tj['tapOutStopsName'].map(data_halte_count.set_index('tapInStopsName')['tapInStops']))

In [30]:
transjakarta.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID          1125
corridorName        1125
direction              0
tapInStops            22
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops         1353
tapOutStopsName     1344
tapOutStopsLat      1344
tapOutStopsLon      1344
stopEndSeq          1344
tapOutTime          1344
payAmount           1007
dtype: int64

## **MAKING DataFrame for LOOKUP condition : data_customer_stasiun**

Because there's only 2000 unique customer inside our data, i suspect that there might be a lot of repeat usage of stations, meaning the same customer tapOut in the same location every time they tapIn from a certain location. this LOOKUP pivot table will be used to prove that conjecture

In [31]:
data_customer_stasiun = pd.pivot_table(data = tj,
               index = ['payCardID','tapInStops', 'tapOutStops'],
               values = 'tapInStopsLat',
               aggfunc= 'count').sort_values(by=['tapInStopsLat'], ascending = False).reset_index().rename(columns={'tapInStopsLat': 'used'})

data_customer_stasiun

payCardID tapInStops tapOutStops  used
0     4997693930906134609    B02234P     B03795P    20
1     4963710913075879668    B03885P     B06067P    20
2     4957958811110735429    B04678P     B05764P    20
3     4941966320527194737     P00237      P00074    20
4     4916430456848946627    B06803P     B05728P    20
...                   ...        ...         ...   ...
4942         501830527583     P00033      P00002     1
4943         501830527583     P00129      P00079     1
4944         501834837947    B00248P      P00016     1
4945         501834837947    B00866P     B04005P     1
4946       30207713127087     P00046      P00239     1

[4947 rows x 4 columns]

In [32]:
data_customer_stasiun.groupby('used')[['payCardID']].count()

payCardID
used           
1          2946
2             1
5             9
6            67
7           324
15            1
16           10
17           38
18          193
19          593
20          765

The conjecture turns out correct, there's a lot of reccuring users of the transjakarta inside our data with the same tapIn and tapOut combination, this means we could predict and fillNA on tapOut based on where the certain customer tapIN

## **FILLING tapOutStops with MAP and Composite_key to fullfill 2 conditions**

Now filling the NA coulnd't be done with just one condition on the MAP, from our previous conjecture, the same customer with the a certain TapIn location would tapOut in the same location as the usually do. this means the MAP that we're using currently have to fullfill 2 conditions of the same customer and the same tapIn instead of only 1 like how it's previously done. To do this we could make a composite key of those 2 conditions where we combine those condition into one inside a single collumns called composite key and then MAP the data with that composite key.

In [33]:
# Making the composite Key
tj['composite_key'] = tj['payCardID'].astype(str) + '_' + tj['tapInStops'].astype(str)
data_customer_stasiun['composite_key'] = data_customer_stasiun['payCardID'].astype(str) + '_' + data_customer_stasiun['tapInStops'].astype(str)

# Deleting Duplicate
data_customer_stasiun.drop_duplicates(subset= 'composite_key', inplace = True)

tj['tapOutStops'] = tj['tapOutStops'].fillna(
    tj['composite_key'].map(data_customer_stasiun.set_index('composite_key')['tapOutStops'])
)

In [34]:
tj.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID          1125
corridorName        1125
direction              0
tapInStops            22
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops          131
tapOutStopsName     1344
tapOutStopsLat      1344
tapOutStopsLon      1344
stopEndSeq          1344
tapOutTime          1344
payAmount           1007
composite_key          0
dtype: int64

## **MAKING DataFrame for LOOKUP condition : tapOutDetails**

We will now be filling all of the missing collumns of the tapOutStops with a single LOOKUP table

In [35]:
tapOutDetails = pd.pivot_table(data = tj,
               index = ['tapOutStops','tapOutStopsName','tapOutStopsLat','tapOutStopsLon','stopEndSeq'],
               values = 'transID',
               aggfunc= 'count').sort_values(by=['transID']).reset_index().rename(columns={'transID': 'used'}).sort_values('used', ascending = False)

tapOutDetails.drop_duplicates(subset='tapOutStops', inplace = True)

In [36]:
# Fill the Stops name
tj['tapOutStopsName'] = tj['tapOutStopsName'].fillna(
    tj['tapOutStops'].map(tapOutDetails.set_index('tapOutStops')['tapOutStopsName']))

In [37]:
# Fill the Stops Latitude
tj['tapOutStopsLat'] = tj['tapOutStopsLat'].fillna(
    tj['tapOutStops'].map(tapOutDetails.set_index('tapOutStops')['tapOutStopsLat']))

In [38]:
# Fill the Stops Longitude
tj['tapOutStopsLon'] = tj['tapOutStopsLon'].fillna(
    tj['tapOutStops'].map(tapOutDetails.set_index('tapOutStops')['tapOutStopsLon']))

In [39]:
# Fill the Stops Sequence
tj['stopEndSeq'] = tj['stopEndSeq'].fillna(
    tj['tapOutStops'].map(tapOutDetails.set_index('tapOutStops')['stopEndSeq']))

In [40]:
tj.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID          1125
corridorName        1125
direction              0
tapInStops            22
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops          131
tapOutStopsName      122
tapOutStopsLat       122
tapOutStopsLon       122
stopEndSeq           122
tapOutTime          1344
payAmount           1007
composite_key          0
dtype: int64

## **MAKING DataFrame for LOOKUP condition : corridor_name_sequence**

Since there is still a lot of corridor ID and Name that is NA, we have to find a new way to fill it. What we can do is we make a LOOKUP table for corridorID with every single avaliable stopsName with their directions and sequence. We have to use direction and sequence because the same Stops or Halte is used by multiple Route, so we have to match a every Stops with the routes directions and sequence. The same Stops will have a different sequence and direction for their routes.

In [41]:
# Making the pivot table
corridor_name_sequence = pd.pivot_table(
    tj,
    index = ['corridorID', 'corridorName', 'direction', 'tapInStops', 'tapInStopsName'],
    values = 'stopStartSeq',
    aggfunc = 'median'
).sort_values(by=['corridorID','direction', 'stopStartSeq']).reset_index()
corridor_name_sequence

corridorID                   corridorName  direction tapInStops  \
0             1                  Blok M - Kota        0.0     P00017   
1             1                  Blok M - Kota        0.0     P00133   
2             1                  Blok M - Kota        0.0     P00012   
3             1                  Blok M - Kota        0.0     P00256   
4             1                  Blok M - Kota        0.0     P00137   
...         ...                            ...        ...        ...   
3805        T21  Palem Semi - Bundaran Senayan        1.0    B00322P   
3806        T21  Palem Semi - Bundaran Senayan        1.0    B06909P   
3807        T21  Palem Semi - Bundaran Senayan        1.0    B05770P   
3808        T21  Palem Semi - Bundaran Senayan        1.0    B00808P   
3809        T21  Palem Semi - Bundaran Senayan        1.0    B02861P   

          tapInStopsName  stopStartSeq  
0                 Blok M           0.0  
1           Masjid Agung           2.0  
2        Bendungan Hilir           6.0  
3             MH Thamrin          11.0  
4                  Monas          13.0  
...                  ...           ...  
3805           DPR MPR 2           6.0  
3806  Slipi Petamburan 3           7.0  
3807     Tol Kb. Jeruk 1           8.0  
3808             Islamic          10.0  
3809        Plaza Europa          11.0  

[3810 rows x 6 columns]

Because as we found previously that the tapIn and tapOut have contains different unique Stops name, we have to include both for out LOOKUP table

In [42]:
corridor_name_sequence2 = pd.pivot_table(
    tj,
    index = ['corridorID', 'corridorName', 'direction', 'tapOutStops', 'tapOutStopsName'],
    values = 'stopEndSeq',
    aggfunc = 'median'
).sort_values(by=['corridorID','direction', 'stopEndSeq']).reset_index()
corridor_name_sequence2

corridorID                   corridorName  direction tapOutStops  \
0             1                  Blok M - Kota        0.0      P00183   
1             1                  Blok M - Kota        0.0      P00010   
2             1                  Blok M - Kota        0.0      P00137   
3             1                  Blok M - Kota        0.0      P00295   
4             1                  Blok M - Kota        0.0      P00139   
...         ...                            ...        ...         ...   
3299        T21  Palem Semi - Bundaran Senayan        1.0     B00808P   
3300        T21  Palem Semi - Bundaran Senayan        1.0     B02861P   
3301        T21  Palem Semi - Bundaran Senayan        1.0     B02557P   
3302        T21  Palem Semi - Bundaran Senayan        1.0     B01440P   
3303        T21  Palem Semi - Bundaran Senayan        1.0     B02556P   

                tapOutStopsName  stopEndSeq  
0              Polda Metro Jaya         5.0  
1     Bank Indonesia Arah Utara        12.0  
2                         Monas        13.0  
3       Mangga Besar Arah Utara        16.0  
4                         Olimo        17.0  
...                         ...         ...  
3299                    Islamic        10.0  
3300               Plaza Europa        11.0  
3301               Palem Semi 2        12.0  
3302     Jln. Palem Jepang Baru        13.0  
3303               Palem Semi 1        14.0  

[3304 rows x 6 columns]

Turns out using LOOKUP table for this process would take a long time and compute, the easier way is to use a dictionary inside a dictionary, the outer dictionary will use the Route or coriddorID and direction as the key, the value is another dictionary with their sequence as key, and the Stops name as values.

In [43]:
# making the corridorID + direction
corridor_name_sequence.insert(2, 'concated_key', corridor_name_sequence['corridorID'].astype(str) + '_' + corridor_name_sequence['direction'].astype(str))

In [44]:
# making the corridorID + direction
corridor_name_sequence2.insert(2, 'concated_key', corridor_name_sequence2['corridorID'].astype(str) + '_' + corridor_name_sequence2['direction'].astype(str))

Turning each LOOKUP table into a dictionary

In [45]:
grouped_directions_in = corridor_name_sequence.groupby('concated_key').apply(
    lambda x: x[['stopStartSeq', 'tapInStopsName']].values.tolist()
)

# Convert to dictionary
grouped_directions_in_dict = grouped_directions_in.to_dict()

In [46]:
grouped_directions_out = corridor_name_sequence2.groupby('concated_key').apply(
    lambda x: x[['stopEndSeq', 'tapOutStopsName']].values.tolist()
)

# Convert to dictionary
grouped_directions_out_dict = grouped_directions_out.to_dict()

Combining the 2 dictionary into a single Dictionary

In [47]:
# Initialize an empty dictionary for the result
combined_dict = {}

# Merge the dictionaries
for key in set(grouped_directions_in_dict) | set(grouped_directions_out_dict):  # Union of all keys in both dictionaries
    combined_values = {}
    if key in grouped_directions_in_dict:
        combined_values.update({item[0]: item[1] for item in grouped_directions_in_dict[key]})  # Convert each list to tuple and add to set
    if key in grouped_directions_out_dict:
        combined_values.update({item[0]: item[1] for item in grouped_directions_out_dict[key]})  # Convert each list to tuple and add to set
    
    # Convert back to list of lists and sort if necessary
    combined_dict[key] = combined_values

combined_dict

{'3C_1.0': {0.0: 'Penjaringan', 1.0: 'Rusun Kapuk Muara'},
 '10H_1.0': {0.0: 'Blok M',
  3.0: 'Bundaran Senayan',
  5.0: 'Slipi Petamburan Arah Utara',
  6.0: 'Slipi Kemanggisan Arah Utara',
  10.0: 'Petojo',
  11.0: 'Pecenongan',
  13.0: 'Pasar Baru',
  14.0: 'Pasar Baru Timur',
  8.0: 'Tomang Mandala',
  15.0: 'Jembatan Merah',
  16.0: 'Gunung Sahari Mangga Dua',
  17.0: 'Pademangan',
  18.0: 'Tanjung Priok'},
 'JAK.60_1.0': {0.0: 'Kemayoran Landas Pacu Timur 2',
  4.0: 'Jln. H. Mawar Sunter',
  6.0: 'Jln. Sunter Jaya VI',
  16.0: 'Wisma SMR',
  18.0: 'SPBU Yos Sudarso',
  19.0: 'Sunter Kelapa Gading 1',
  23.0: 'Ruko Gading Kirana',
  25.0: 'Ruko Gading Kirana 3',
  29.0: 'Sbr. Jakarta Taipei School',
  34.0: 'Jln. Kelapa Hybrida IV',
  37.0: 'Pusat Pelayanan Warga Summarecon',
  38.0: 'Singapore International School',
  24.0: 'Ruko Gading Kirana 2',
  28.0: 'Mitra Gading Villa',
  30.0: 'Sbr. BPK Penabur Kelapa Gading',
  33.0: 'Jln. Kelapa Hybrida 10',
  35.0: 'Jln. Gading Ayu I',

Matching the dictionary into the DataFrame

In [48]:
def find_combined_key(row, combined_dict):
    for outer_key, inner_dict in combined_dict.items():
        if row['stopStartSeq'] in inner_dict and inner_dict[row['stopStartSeq'] == row['tapInStopsName']]:
            return outer_key
        return row['corridorID']

tj['corridorID'] = tj.apply(find_combined_key, axis=1, combined_dict=combined_dict)

In [49]:
tj.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID           981
corridorName        1125
direction              0
tapInStops            22
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops          131
tapOutStopsName      122
tapOutStopsLat       122
tapOutStopsLon       122
stopEndSeq           122
tapOutTime          1344
payAmount           1007
composite_key          0
dtype: int64

Rematching the corridorName with the newly found corridorID

In [50]:
# Making a pivot table, since the corridorName has lower count, there should be a corridorName with multiple corridorID
data_stasiun_count = pd.pivot_table(data = tj,
               index = ['corridorName','corridorID'],
               values = 'tapInStopsLat',
               aggfunc= 'count').sort_values(by=['tapInStopsLat'], ascending=False).reset_index().rename(columns={'tapInStopsLat': 'Used'})
data_stasiun_count.drop_duplicates(subset='corridorID', inplace=True)

tj['corridorName'] = tj['corridorName'].fillna(
    tj['corridorID'].str.upper().map(data_stasiun_count.set_index('corridorID')['corridorName'])
)

In [51]:
tj.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID           981
corridorName         981
direction              0
tapInStops            22
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops          131
tapOutStopsName      122
tapOutStopsLat       122
tapOutStopsLon       122
stopEndSeq           122
tapOutTime          1344
payAmount           1007
composite_key          0
dtype: int64

## **Data Distribution & Outliers**

Since all of the data presented in our data source is categorical nominal, we do not need to check the data distribution nor do we have to do normality test to check whether the data is normally distributed or not. Because of that, we also do not have outliers. The only numeric data we have is the Pay Amount, but even then it only contains numeric discreet values.

In [52]:
tj[tj['payAmount'] == 0]

transID            payCardID payCardBank  \
6      HEMW326B9N91TV      213155822653833      emoney   
7      XTKE052E5E87LN     3587341902618993         dki   
14     LZFD368Q3E81HJ        4056836188316      online   
19     ZYRL615G4C05RD     5430018591726410         dki   
21     ZJVI960B9I42DA      343851798729291      emoney   
...               ...                  ...         ...   
37891  XQDZ821C9G88JD     2712823212983860         dki   
37892  YHJT665I6G08OS        4775206940093      online   
37893  GNET512K3A93CA     3500965207195341         dki   
37894  ZXVG342K6T27GU  4475487986105118550      brizzi   
37898  RGVK175U2U98UV      377840859133591      emoney   

                      payCardName payCardSex  payCardBirthDate corridorID  \
6           R. Devi Hariyah, S.T.          M              1974     JAK.18   
7                 Darmanto Rajasa          F              1991        NaN   
14                   Nadine Utami          M              1975     JAK.46   
19                  Karna Waskita          M              1981     JAK.88   
21                   Garang Utama          F              1997     JAK.16   
...                           ...        ...               ...        ...   
37891    drg. Maras Wibowo, S.Ked          F              1966     JAK.30   
37892       Cahyo Sudiati, M.Farm          M              1979     JAK.52   
37893            Latika Salahudin          F              2000     JAK.80   
37894  Tgk. Dipa Purnawati, S.E.I          F              2012     JAK.39   
37898                Ana Agustina          M              1976     JAK.13   

                                    corridorName  direction tapInStops  \
6                            Kalibata - Kuningan        0.0    B00243P   
7                                            NaN        1.0    B03416P   
14                      Pasar Minggu - Jagakarsa        1.0    B00608P   
19          Terminal Tanjung Priok - Ancol Barat        0.0    B00722P   
21                            Cililitan - Condet        0.0    B02144P   
...                                          ...        ...        ...   
37891                   Grogol - Meruya via Roxy        1.0    B01209P   
37892  Terminal Kalideres - Terminal Muara Angke        1.0    B01087P   
37893                  Rawa Buaya - Rawa Kompeni        0.0    B02485P   
37894                   Kalimalang - Duren Sawit        0.0    B04489P   
37898                Tanah Abang - Jembatan Lima        1.0    B02505P   

                       tapInStopsName  tapInStopsLat  tapInStopsLon  \
6                     Cervino Village      -6.224355      106.85165   
7                  SDN Pondok Labu 11      -6.313269      106.80221   
14     Gelanggang Remaja Pasar Minggu      -6.284615      106.83787   
19     Gunung Sahari Mangga Dua Timur      -6.137032      106.83254   
21             Lippo Mall Kramat Jati      -6.270835      106.86741   
...                               ...            ...            ...   
37891     Jln. Kartika Meruya Selatan      -6.203099      106.73673   
37892        Jln. Gunung Galunggung 1      -6.142286      106.73712   
37893                         MTsN 37      -6.097910      106.70237   
37894                   Sbr. SMPN 252      -6.239289      106.94229   
37898                  Museum Textile      -6.188656      106.80954   

       stopStartSeq            tapInTime tapOutStops  \
6                25  2023-04-03 06:58:21     B04699P   
7                 2  2023-04-03 06:44:24     B00899P   
14                5  2023-04-03 05:38:44     B01062P   
19               20  2023-04-03 06:43:49     B05989P   
21                4  2023-04-03 05:08:36     B01393P   
...             ...                  ...         ...   
37891            14  2023-04-19 16:31:09     B04071P   
37892            29  2023-04-28 16:38:56     B03207P   
37893            39  2023-04-19 08:16:10     B03053P   
37894            27  2023-04-14 11:36:12     B01624P   
37898            33  2023-04-15 19

Almost half of the data have a payAmount of 0, to get a price of Rp. 0, Indonesia citizens have to fulfill one of 15 conditions including, civil worker, elderly, disabled, etc.
for the one with the price of Rp. 20.000, that is for a different type of transjakarta bus, called RoyalTrans which consist of only less than 5% of data

In [53]:
tj[tj['payAmount'] == 20000]

transID         payCardID payCardBank  \
5      DDES630K2F80KC  2251412124634980         dki   
8      OIHS248V7S72EB  6510013988638519         dki   
60     RHUT145C1U25ZY     4290559479079      online   
70     OTGH855B8A29BP  3549158005951908         dki   
117    YTOA058V9U73VN     4709994160903      online   
...               ...               ...         ...   
37854  AMQA420I2P82KP  3553882732550120         dki   
37878  FFUS474T7Q54QF    36067042340502         bni   
37879  LVSU593P5B50VF    30480183587078         bni   
37888  KWQA944F9O89IN  6585885975007109         dki   
37897  YXPP627N4G95HO   213159426675861      emoney   

                         payCardName payCardSex  payCardBirthDate corridorID  \
5                      Tirta Siregar          F              1993         1T   
8             Cagak Maheswara, S.Sos          F              1992     3C_1.0   
60                      Lutfan Hakim          M              2011        T21   
70         Mustofa Wijayanti, M.Farm          M              1981         1T   
117    Cut Gabriella Haryanti, S.Sos          M              1974        T21   
...                              ...        ...               ...        ...   
37854                   Siti Rahimah          M              1994         6P   
37878                 Jatmiko Mansur          F              1965        D32   
37879   Hj. Julia Nainggolan, M.Kom.          M              1990        S12   
37888                Hardana Gunarto          F              1982        B14   
37897         drg. Zahra Nashiruddin          F              1976         1T   

                        corridorName  direction tapInStops   tapInStopsName  \
5               Cibubur - Balai Kota        0.0    B00127P         Benhil 3   
8            Bekasi Barat - Kuningan        1.0    B00795P         Imperium   
60     Palem Semi - Bundaran Senayan        1.0    B00589P      GBK Pintu 7   
70              Cibubur - Balai Kota        0.0    B02408P     Menara Astra   
117    Palem Semi - Bundaran Senayan        1.0    B00808P          Islamic   
...                              ...        ...        ...              ...   
37854             Cibubur - Kuningan        1.0    B00596P   Gedung Granadi   
37878      Cinere - Bundaran Senayan        0.0    B06642P      Mega Cinere   
37879        BSD Serpong - Fatmawati        1.0    B04581P  Sbr. Versailles   
37888        Bekasi Barat - Kuningan        0.0    B00664P   GOR Sumantri 2   
37897           Cibubur - Balai Kota        1.0    B02873P    Plaza Sentral   

       tapInStopsLat  tapInStopsLon  stopStartSeq            tapInTime  \
5          -6.216010      106.81632             3  2023-04-03 05:08:01   
8          -6.210363      106.83059             1  2023-04-03 06:01:44   
60         -6.222095      106.80861             3  2023-04-03 05:43:48   
70         -6.207361      106.82190             7  2023-04-03 05:57:55   
117        -6.228521      106.61473            10  2023-04-03 06:37:27   
...              ...            ...           ...                  ...   
37854      -6.231311      106.83301            11  2023-04-03 17:56:55   
37878      -6.333340      106.78282             4  2023-04-16 07:59:08   
37879      -6.308522      106.67339             3  2023-04-07 09:28:01   
37888      -6.221038      106.83189            11  2023-04-05 09:30:00   
37897      -6.216247      106.81676            12  2023-04-04 10:29:47   

      tapOutStops  tapOutStopsName  tapOutStopsLat  tapOutStopsLon  \
5         B00694P      Grand Sahid       -6.210975       106.82059   
8         B01853P     Kayuringin 2       -6.245948       106.99247   
60        B02556P     Palem Semi 1       -6.218812       106.61807   
70        B02420P   Menara Thamrin       -6.185275       106.82291   
117       B02557P     Palem Semi 2       -6.220361       106.61649   
...           ...              ...             ...             ...   
37854     B00226P  Buperta Cibubur       -6.370321       106.896

for the remaining data which is NA, i will fill it using the normal price of 3500 rupiah

In [54]:
tj['payAmount'].fillna(3500, inplace=True)

# **Data Handling of Time**

In the data there is 2 columns containing Time information, for example the value inside 1 cell would be like this: 2023-04-03 05:08:01.

For better data analysis and finding pattern, this kind of format is not manageable and doesn't give much insight. First of all, all the data inside our DataFrame is taken on transaction that happens in april 2023 so we dont need the month or the year of the format. we only need the dates and hour. Secondly it will be better to analyze the data not just by dates, but also by what day it is because there is usually a pattern showing how the same day have the same amount of users. that means there will be a new columns of day. And lastly, the hour and minutes of the time format doesn't have much function if we don't compare it from the first time the customer tapIn until the customer tapOut, in other words, durations of the Transjakarta usage. so there will be another columns of duration.

### **Data Cleaning**

Before we start handling Time, i would like to clean the data even more. After a thorough cleaning, there is still a few NA inside corridorName, ID, and tapOutDetails. I have used every possible way to my current knowledge to fill the NA and can't find any other way to fill the remaining missing data except for the columns of TapOutTime. Because of that, I will be using DROPNA for the remaining columns missing data.

In [55]:
tj.dropna(subset='corridorID', inplace = True)

In [56]:
tj.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID             0
corridorName           0
direction              0
tapInStops            22
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops          129
tapOutStopsName      120
tapOutStopsLat       120
tapOutStopsLon       120
stopEndSeq           120
tapOutTime          1305
payAmount              0
composite_key          0
dtype: int64

In [57]:
tj.dropna(subset='tapOutStops', inplace = True)
tj.dropna(subset='tapInStops', inplace = True)

In [58]:
tj.isna().sum()

transID                0
payCardID              0
payCardBank            0
payCardName            0
payCardSex             0
payCardBirthDate       0
corridorID             0
corridorName           0
direction              0
tapInStops             0
tapInStopsName         0
tapInStopsLat          0
tapInStopsLon          0
stopStartSeq           0
tapInTime              0
tapOutStops            0
tapOutStopsName        0
tapOutStopsLat         0
tapOutStopsLon         0
stopEndSeq             0
tapOutTime          1185
payAmount              0
composite_key          0
dtype: int64

we will now convert the stopEndSeq into an integer, we can't do it previously because of the missing values

In [59]:
tj['stopEndSeq'] = tj['stopEndSeq'].astype(int)

I will also drop the composite_key columns because we won't be using it anymore

In [60]:
tj.drop(columns = 'composite_key', inplace = True)

### **Changing type to DateTime**

In [61]:
# Changing the dtype
tj['tapInTime'] = pd.to_datetime(tj['tapInTime'])
tj['tapOutTime'] = pd.to_datetime(tj['tapOutTime'])

# Create new columns
tj['date'] = tj['tapInTime'].dt.strftime('%d')  # Extract the day of the month as "03"
tj['day'] = tj['tapInTime'].dt.strftime('%A')  # Extract the day name like "Monday"
tj['tapInHour'] = tj['tapInTime'].dt.strftime('%H:%M:%S')  # Extract the time as "05:08:01"
tj['tapOutHour'] = tj['tapOutTime'].dt.strftime('%H:%M:%S')

In [62]:
# Calculate the duration
tj['duration'] = tj['tapOutTime'] - tj['tapInTime']

# Format the duration to show only the time part (HH:MM:SS)
# tj['duration'] = tj['duration'].apply(lambda x: str(x).split(' ')[-1])

In [63]:
tj[['date', 'day', 'tapInTime','tapOutTime','duration']]

date        day           tapInTime          tapOutTime        duration
0       03     Monday 2023-04-03 05:21:44 2023-04-03 06:00:53 0 days 00:39:09
1       03     Monday 2023-04-03 05:42:44 2023-04-03 06:40:01 0 days 00:57:17
2       03     Monday 2023-04-03 05:59:06 2023-04-03 06:50:55 0 days 00:51:49
3       03     Monday 2023-04-03 05:44:51 2023-04-03 06:28:16 0 days 00:43:25
4       03     Monday 2023-04-03 06:17:35 2023-04-03 06:57:03 0 days 00:39:28
...    ...        ...                 ...                 ...             ...
37895   21     Friday 2023-04-21 18:18:37 2023-04-21 19:55:49 0 days 01:37:12
37896   18    Tuesday 2023-04-18 21:52:31 2023-04-18 22:28:22 0 days 00:35:51
37897   04    Tuesday 2023-04-04 10:29:47 2023-04-04 13:27:25 0 days 02:57:38
37898   15   Saturday 2023-04-15 19:59:26 2023-04-15 20:27:50 0 days 00:28:24
37899   12  Wednesday 2023-04-12 21:08:12 2023-04-12 21:34:53 0 days 00:26:41

[36768 rows x 5 columns]

### **Filling the remaining missing data**

Now we can try to fill the remaining missing data in the columns of tapOutTime, we can do it by averaging how long does the bus take to go from one stops to the next stops within the same route. then we can fill the tapOutTime just by adding the average time it takes to go for a single stops, and multiplied it to the total number of stops the users has to go through.

In [64]:
tj['stopsPassed'] = tj['stopEndSeq'] - tj['stopStartSeq']

In [65]:
# Making the pivot table to group by corridorID and average by duration per stopsPassed
avgDuration = pd.pivot_table(data = tj,
               index = ['corridorID','stopsPassed'],
               values = 'duration',
               aggfunc= 'mean')

avgDuration.reset_index(inplace=True)

avgDuration['averagePerStops'] = avgDuration['duration'] / avgDuration['stopsPassed']

avgDuration

corridorID  stopsPassed                  duration  \
0             1            1    0 days 01:07:16.500000   
1             1            2 0 days 01:17:42.794871794   
2             1            3 0 days 01:19:18.666666666   
3             1            4           0 days 01:19:58   
4             1            5 0 days 00:55:50.736842105   
...         ...          ...                       ...   
2369        T21            3 0 days 01:26:07.428571428   
2370        T21            4    0 days 00:58:53.650000   
2371        T21            5    0 days 01:16:50.600000   
2372        T21            6           0 days 01:51:07   
2373        T21           11 0 days 00:35:28.684210526   

               averagePerStops  
0       0 days 01:07:16.500000  
1    0 days 00:38:51.397435897  
2    0 days 00:26:26.222222222  
3       0 days 00:19:59.500000  
4    0 days 00:11:10.147368421  
...                        ...  
2369 0 days 00:28:42.476190476  
2370    0 days 00:14:43.412500  
2371    0 days 00:15:22.120000  
2372 0 days 00:18:31.166666666  
2373 0 days 00:03:13.516746411  

[2374 rows x 4 columns]

In [66]:
# finding the total average of each corridorID
avgTotal = pd.pivot_table(data = avgDuration,
               index = 'corridorID',
               values = 'averagePerStops',
               aggfunc= 'mean').reset_index()

# Deleting the fractions of seconds & days
avgTotal['averagePerStops'] = avgTotal['averagePerStops'].astype(str).str.split('.').str[0]
#avgTotal['averagePerStops'] = avgTotal['averagePerStops'].apply(lambda x: str(x).split(' ')[-1])
avgTotal['averagePerStops'] = pd.to_timedelta(avgTotal['averagePerStops'])
# avgTotal['averagePerStops'] = pd.to_datetime(avgTotal['averagePerStops'])

avgTotal

corridorID averagePerStops
0            1 0 days 00:19:27
1           10 0 days 00:24:03
2          10A 0 days 00:34:18
3          10B 0 days 00:50:09
4          10D 0 days 00:21:40
..         ...             ...
212        S21 0 days 00:14:20
213        S22 0 days 00:27:40
214        S31 0 days 00:38:59
215        T11 0 days 00:16:19
216        T21 0 days 00:25:17

[217 rows x 2 columns]

In [67]:
# Merge df into tj to get the average duration for each corridorID
tj = tj.merge(avgTotal, on='corridorID', how='left')

In [68]:
tj.loc[6]

transID                           HEMW326B9N91TV
payCardID                        213155822653833
payCardBank                               emoney
payCardName                R. Devi Hariyah, S.T.
payCardSex                                     M
payCardBirthDate                            1974
corridorID                                JAK.18
corridorName                 Kalibata - Kuningan
direction                                    0.0
tapInStops                               B00243P
tapInStopsName                   Cervino Village
tapInStopsLat                          -6.224355
tapInStopsLon                          106.85165
stopStartSeq                                  25
tapInTime                    2023-04-03 06:58:21
tapOutStops                              B04699P
tapOutStopsName     Simpang Al Barkah Dr Saharjo
tapOutStopsLat                         -6.220395
tapOutStopsLon                         106.84653
stopEndSeq                                    40
tapOutTime          

In [69]:
# Calculate the filled values for 'time' column
tj['tapOutHour'] = tj.apply(lambda row: row['averagePerStops'] * row['stopsPassed'] if pd.isna(row['tapOutHour']) else row['tapOutHour'], axis=1)

tj['tapOutTime'] = tj.apply(lambda row: row['tapInTime'] + row['tapOutHour'] if pd.isna(row['tapOutTime']) else row['tapOutTime'], axis=1)

In [70]:
tj['tapInTime'] = pd.to_datetime(tj['tapInTime'], format='0 days %H:%M:%S')
tj['tapOutTime'] = pd.to_datetime(tj['tapOutTime'], format='0 days %H:%M:%S')

# Calculate duration
tj['duration'] = tj['tapOutTime'] - tj['tapInTime']
tj['tapOutHour'] = tj['tapOutTime'].dt.strftime('%H:%M:%S')

In [71]:
tj['duration'] = tj['duration'].dt.total_seconds()
tj['duration'] = tj['duration'] / 60

In [72]:
tj.isna().sum()

transID             0
payCardID           0
payCardBank         0
payCardName         0
payCardSex          0
payCardBirthDate    0
corridorID          0
corridorName        0
direction           0
tapInStops          0
tapInStopsName      0
tapInStopsLat       0
tapInStopsLon       0
stopStartSeq        0
tapInTime           0
tapOutStops         0
tapOutStopsName     0
tapOutStopsLat      0
tapOutStopsLon      0
stopEndSeq          0
tapOutTime          0
payAmount           0
date                0
day                 0
tapInHour           0
tapOutHour          0
duration            0
stopsPassed         0
averagePerStops     0
dtype: int64

In [73]:
tj

transID            payCardID payCardBank  \
0      EIIW227B8L34VB      180062659848800      emoney   
1      LGXO740D2N47GZ     4885331907664776         dki   
2      DJWR385V2U57TO     4996225095064169         dki   
3      JTUZ800U7C86EH         639099174703       flazz   
4      VMLO535V7F95NJ         570928206772       flazz   
...               ...                  ...         ...   
36763  ZWEC949B8Q87QG  4685818286724028395      brizzi   
36764  YHHK837P6Y95GN     6502902290603767         dki   
36765  YXPP627N4G95HO      213159426675861      emoney   
36766  RGVK175U2U98UV      377840859133591      emoney   
36767  FMZZ963S4B68ZP         501862539795       flazz   

                  payCardName payCardSex  payCardBirthDate corridorID  \
0              Bajragin Usada          M              2008          5   
1                Gandi Widodo          F              1997         6C   
2               Emong Wastuti          F              1992        R1A   
3                Surya Wacana          F              1978        11D   
4             Embuh Mardhiyah          M              1982         12   
...                       ...        ...               ...        ...   
36763         Kamila Mahendra          F              2004         6B   
36764            Titi Siregar          M              1974     3C_1.0   
36765  drg. Zahra Nashiruddin          F              1976         1T   
36766            Ana Agustina          M              1976     JAK.13   
36767       drg. Leo Najmudin          F              1985         13   

                                   corridorName  direction tapInStops  \
0                         Matraman Baru - Ancol        1.0     P00142   
1      Stasiun Tebet - Karet via Patra Kuningan        0.0    B01963P   
2                            Pantai Maju - Kota        0.0    B00499P   
3           Pulo Gebang - Pulo Gadung 2 via PIK        0.0    B05587P   
4                         Tanjung Priok - Pluit        0.0     P00239   
...                                         ...        ...        ...   
36763         Ragunan - MH Thamrin via Semanggi        1.0     P00261   
36764                    Pinang Ranti - Pramuka        1.0     P00064   
36765                      Cibubur - Balai Kota        1.0    B02873P   
36766               Tanah Abang - Jembatan Lima        1.0    B02505P   
36767                         Ciledug - Tendean        0.0     P00001   

               tapInStopsName  tapInStopsLat  tapInStopsLon  stopStartSeq  \
0                   Pal Putih      -6.184631      106.84402             7   
1                  Kemenkes 2      -6.228700      106.83302            13   
2                Gg. Kunir II      -6.133132      106.81435            38   
3                Taman Elok 1      -6.195743      106.93526            23   
4      Sunter Boulevard Barat      -6.149650      106.88900             5   
...                       ...            ...            ...           ...   
36763                  Tosari      -6.196892      106.82309             2   
36764       Garuda Taman Mini      -6.290154      106.88116             1   
36765           Plaza Sentral      -6.216247      106.81676            12   
36766          Museum Textile      -6.188656      106.80954            33   
36767              Adam Malik      -6.236466      106.74786             2   

                tapInTime tapOutStops        tapOutStopsName  tapOutStopsLat  \
0     2023-04-03 05:21:44      P00253                Tegalan       -6.203101   
1     2023-04-03 05:42:44     B03307P    Sampoerna Strategic       -6.217152   
2     2023-04-03 05:59:06     B04962P  Simpang Kunir Kemukus       -6.133731   
3     2023-04-03 05:44:51     B03090P      Raya Penggilingan       -6.183068   
4     2023-04-03 06:17:35      P00098       Kali Besar Barat       -6.135355   
...                   ...         ...                    ...             ...   
36763 2023-04-21 18:18:37      P00228                 SMK 57       -6.290967  

In [74]:
tj.to_csv('transjakarta_clean.csv', index=False)

# **3. Data Analysis**
*next file...*